### PR TITLE
Day19 Ex01 Music

### PR DESCRIPTION
**Pull request by at-hangtran**
#### Related issues (Optional)
Task  #100 

#### Description (Required)
- [x] Play music with recyclerView and control buttons
- [x] Show notification when screen off
- [x] Update data in notification

### CODE REVIEW

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        musicViewHolder.mTvSinger.setText(mSongs.get(position).getSinger());
+        musicViewHolder.mImgSong.setImageResource(mSongs.get(position).getImage());
+    }
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
+
+    private class MusicViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            itemView.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    listener.onItemClick(song, position);
+                }
+            });
+        }
+    }
+
+    interface OnItemClickListener {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/SongFragment.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        musicRecyclerView.setAdapter(adapter);
+
+        return view;
+    }
+
+    private void getSong(Song song, int position) {
+        mListener.onGetSong(song, position);
+    }
+
+    interface OnGetSongListener {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/SongFragment.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicActivity.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+//        mSongs.add(new Song("Stay", "Zedd, Alessia Cara", "http://api.mp3.zing.vn/api/mobile/source/song/LGJGTLGNQJQJLQDTLDJTDGLG", R.drawable.img_stay));
+//        mSongs.add(new Song("Believer", "Imagine Dragons", "http://api.mp3.zing.vn/api/mobile/source/song/LGJGTLGNQJDQQLVTLDJTDGLG", R.drawable.img_bliever));
+//        mSongs.add(new Song("Issues", "Julia Michaels", "http://api.mp3.zing.vn/api/mobile/source/song/LGJGTLGNQJLDXGDTLDJTDGLG", R.drawable.img_issue));
+        SongManager songManager = new SongManager();
+        mSongs.addAll(songManager.getListSongOffline(this));
+    }
+
+    private void initView() {
+        mImgBtnPlay = (ImageButton) findViewById(R.id.imgBtnPlay);
+        ImageButton imgBtnNext = (ImageButton) findViewById(R.id.imgBtnNext);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicActivity.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'R.id.imgBtnNext' due to existing static import 'vn.asiantech.internship.R.id.imgBtnNext'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/vn/asiantech/internship/music/MusicActivity.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        ImageButton imgBtnPrev = (ImageButton) findViewById(R.id.imgBtnPrev);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicActivity.java]: WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'R.id.imgBtnPrev' due to existing static import 'vn.asiantech.internship.R.id.imgBtnPrev'
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                                mCurrentSongIndex = 0;
+                            }
+                        }
+                    }
+                    break;
+                case "PREVIOUS":
+                    if (mMediaPlayer != null) {
+                        final Intent playpreviousIntent = new Intent(Action.PREVIOUS.getValue());
+                        sendBroadcast(playpreviousIntent);
+                        if (mMediaPlayer.isPlaying() && mMediaPlayer != null) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: WARNING: [MisplacedNullCheck] The null check here is misplaced; if the variable is null there will be a NullPointerException
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/basic.html#MisplacedNullCheck

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        final Intent updateSongNameIntent = new Intent(Action.UPDATE.getValue());
+        updateSongNameIntent.putExtra("songName", mSongs.get(mCurrentSongIndex).getName());
+        sendBroadcast(updateSongNameIntent);
+        try {
+            mMediaPlayer = new MediaPlayer();
+            mMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
+            mMediaPlayer.setDataSource(mSongs.get(position).getUrl());
+            mMediaPlayer.prepare();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
+
+    /*
+     * Used to register for song
+     */
+    private class MusicViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                    listener.onItemClick(song, position);
+                }
+            });
+        }
+    }
+
+    /*
+     * Used to get song and position in recyclerView
+     */
+    interface OnItemClickListener {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/SongsFragment.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private void getSong(Song song, int position) {
+        mListener.onGetSong(position);
+    }
+
+    /*
+     * Used to get song and positon when click to song list
+     */
+    interface OnGetSongListener {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/SongsFragment.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+                if (mMediaPlayer != null) {
+                    mMediaPlayer.seekTo(intent.getIntExtra("chooseTime", 0));
+                }
+            }
+        }
+    };
+    private final NotificationBroadCast mNotificationBroadCast = new NotificationBroadCast();
+
+    @Override
+    public void onCreate() {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: WARNING: [CallSuperFirst] super should be called at the start of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperFirst

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        if (mMediaPlayer != null) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+            try {
+                mMediaPlayer.release();
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java (hunk 9 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/SongsFragment.java (hunk 10 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/SongsFragment.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        if (mMediaPlayer != null) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/music/MusicActivity.java [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+                prevIntent.setAction(Action.PREVIOUS.getValue());
+                startService(prevIntent);
+                break;
+            case R.id.imgBtnShuffle:
+                Intent shuffleIntent = new Intent(MusicActivity.this, MusicService.class);
+                shuffleIntent.setAction(Action.SHUFFEL.getValue());
+                startService(shuffleIntent);
+                break;
+            case R.id.imgBtnReplay:
+                Intent autoIntent = new Intent(MusicActivity.this, MusicService.class);
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicActivity.java]: please change `autoIntent ` to `replayIntent `

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+ * @author at-HangTran
+ * @version 1.0
+ * @since 2017-7-1
+ */
+public class MusicService extends Service {
+    private static final String TAG = MusicService.class.getSimpleName();
+    private MediaPlayer mMediaPlayer;
+    private List<Song> mSongs = new ArrayList<>();
+    private CountDownTimer mCountDownTimer;
+    private int mCurrentSongIndex = 0;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: remove ` = 0`

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        IntentFilter screenStateFilter = new IntentFilter(Intent.ACTION_SCREEN_OFF);
+        registerReceiver(mNotificationBroadCast, screenStateFilter);
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        IntentFilter mStartFilter = new IntentFilter(Action.SEEKTO.getValue());
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: `mStartFilter` to `startFilter`

File: app/src/main/java/vn/asiantech/internship/music/MusicService.java [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+            } else if (TextUtils.equals(action, Action.AUTONEXT.getValue())) {
+                autoNext();
+            } else if (TextUtils.equals(action, Action.CHOOSESONGFROMLIST.getValue())) {
+                mSongs = intent.getParcelableArrayListExtra("songs");
+                mCurrentSongIndex = intent.getIntExtra("currentSong", 0);
+                playSongWithPosition(mCurrentSongIndex);
+            } else if (TextUtils.equals(action, Action.CANCEL.getValue())) {
+                stopSelf();
+                stopForeground(true);
+                NotificationManager mNotificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicService.java]: `mNotificationManager ` to `notificationManager `

File: app/src/main/java/vn/asiantech/internship/music/SongManager.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.util.ArrayList;
+
+import vn.asiantech.internship.R;
+
+/**
+ * Author AsianTech Inc.
+ * Created by sony on 04/07/2017.
+ */
+
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/SongManager.java]: remove this break line

File: app/src/main/res/layout/activity_music.xml [hunk 20]
```diff
@@ -0,0 +1,97 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="@drawable/img_background">
+
+    <android.support.v4.view.ViewPager
+        android:id="@+id/musicViewPager"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/activity_music.xml]: `musicViewPager` to `viewPagerMusic`

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
+
+    /*
+     * Used to register for song
+     */
+    private class MusicViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+                    listener.onItemClick(song, position);
+                }
+            });
+        }
+    }
+
+    /*
+     * Used to get song and position in recyclerView
+     */
+    interface OnItemClickListener {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicRecyclerViewAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/SongsFragment.java [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    private void getSong(int position) {
+        mListener.onGetSong(position);
+    }
+
+    /*
+     * Used to get song and position when click to song list
+     */
+    interface OnGetSongListener {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/SongsFragment.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/music/MusicActivity.java [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+
+        mImgBtnPlay.setOnClickListener(this);
+        imgBtnNext.setOnClickListener(this);
+        imgBtnPrevious.setOnClickListener(this);
+        mImgBtnShuffle.setOnClickListener(this);
+        mImgBtnReplay.setOnClickListener(this);
+    }
+
+    private void registerBroadcastReceiver() {
+        IntentFilter mStartFilter = new IntentFilter();
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicActivity.java]: `mStartFilter` to `startFilter`

File: app/src/main/java/vn/asiantech/internship/music/MusicTime.java [hunk 25]
```diff
@@ -0,0 +1,29 @@
+package vn.asiantech.internship.music;
+
+/**
+ * Used to handle time that is displayed on textView for seekbar.
+ *
+ * @author at-HangTran
+ * @version 1.0
+ * @since 2017-7-1
+ */
+class MusicTime {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicTime.java]: add `final` for this class.
class name to `MusicUtil`.
create `private constructor` then change `milliSecondsToTimer` to `public static` method

File: app/src/main/java/vn/asiantech/internship/music/Song.java [hunk 26]
```diff
@@ -0,0 +1,90 @@
+package vn.asiantech.internship.music;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/Song.java]: remove break line

File: app/src/main/res/layout/custom_notification.xml [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+        <ProgressBar
+            android:id="@+id/progressBar"
+            style="?android:attr/progressBarStyleHorizontal"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:background="@color/colorWhite" />
+    </LinearLayout>
+
+    <ImageView
+        android:id="@+id/imgCancle"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/custom_notification.xml]: `imgCancle` to `imgCancel`

File: app/src/main/java/vn/asiantech/internship/music/MusicActivity.java [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            ActivityCompat.requestPermissions(MusicActivity.this, new String[]{
+                    Manifest.permission.READ_EXTERNAL_STORAGE}, 1);
+        } else {
+            showApp();
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/music/MusicActivity.java]: ERROR: Array brackets at illegal position.

**Comment by at-vinhhuynh** [comment]
@at-hangtran 
Please fix CI comment first!

File: app/src/main/java/vn/asiantech/internship/music/MusicActivity.java (hunk 24 above)
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/music/MusicActivity.java]: @at-hangtran 
Please fix thís comment

**Comment by at-tienhoang** [comment]
LGTM

**Pull request merged**